### PR TITLE
Validate CTS XML output against the schema

### PR DESCRIPTION
**Pull request by clumens**
This covers some output from crm_mon and crm_resource.  We can add more as we go along.  It will fail until the fix for the crm_resource digests stuff gets merged into master.

### CODE REVIEW

**Comment by kgaillot** [comment]
master should have everything now

File: cts/cts-cli.in [hunk 1]
```diff
@@ -119,15 +123,35 @@ function _test_assert() {
     printf "* Running: $app - $desc\n" 1>&2
 
     printf "=#=#=#= Begin test: $desc =#=#=#=\n"
-    eval $VALGRIND_CMD $cmd 2>&1
+
+    export outfile=$(mktemp ${TMPDIR:-/tmp}/cts-cli.output.XXXXXXXXXX)
+    eval $cmd 2>&1 | tee $outfile
```
Comment (kgaillot) [inline cts/cts-cli.in:128]: Did you mean to kill valgrind here?

**Review by kgaillot — COMMENTED** [review]
looks good except one thing

File: cts/cts-cli.in (hunk 1 above)
Comment (clumens) [reply cts/cts-cli.in]: Nope, that must have been an artifact of when I was first working on the patch.  Fixed.

**Comment by kgaillot** [comment]
Oh right, we don't depend on xmllint, and travis only installs it for the MAINT_EXTRA=1 case. I suppose we could make the validation additionally depend on whether xmllint is available (and install it unconditionally in travis).

**Comment by clumens** [comment]
> Oh right, we don't depend on xmllint, and travis only installs it for the MAINT_EXTRA=1 case. I suppose we could make the validation additionally depend on whether xmllint is available (and install it unconditionally in travis).

Ah, whoops.  I'll fix that up in the morning.

**Review by kgaillot — COMMENTED** [review]
Might as well add xmllint to optional testing dependencies in INSTALL

File: cts/cts-cli.in [hunk 2]
```diff
@@ -119,15 +123,35 @@ function _test_assert() {
...
 
     if [ x$cib != x0 ]; then
         printf "=#=#=#= Current cib after: $desc =#=#=#=\n"
         CIB_user=root cibadmin -Q
     fi
 
-    printf "=#=#=#= End test: $desc - $(crm_error --exit $rc) ($rc) =#=#=#=\n"
+    # Do not validate if running under valgrind, even if told to do so.  Valgrind
+    # will output a lot more stuff that is not XML, so it wouldn't validate anyway.
+    if [ x$validate = x1 -a "$VALGRIND_CMD" = "" -a $rc = 0 -a -x /usr/bin/xmllint ]; then
```
[the remark this answers is not in the document]
Comment (clumens) [reply cts/cts-cli.in]: Fixed, along with some other things shellcheck was complaining about in this area.

File: cts/cts-cli.in [hunk 3]
```diff
@@ -119,15 +124,35 @@ function _test_assert() {
...
 
     if [ x$cib != x0 ]; then
         printf "=#=#=#= Current cib after: $desc =#=#=#=\n"
         CIB_user=root cibadmin -Q
     fi
 
-    printf "=#=#=#= End test: $desc - $(crm_error --exit $rc) ($rc) =#=#=#=\n"
+    # Do not validate if running under valgrind, even if told to do so.  Valgrind
+    # will output a lot more stuff that is not XML, so it wouldn't validate anyway.
+    if [ "$validate" = "1" ] && [ "$VALGRIND_CMD" = "" ] && [ $rc = 0 ] && [ "$validate" -eq 1 ]; then
```
Comment (kgaillot) [inline cts/cts-cli.in:139]: You might think a little more about your variable names :)

File: cts/cts-cli.in [hunk 4]
```diff
@@ -1718,6 +1743,12 @@ for t in $tests; do
     esac
 done
 
+if [ -x /usr/bin/xmllint ]; then
+    validate=1
+else
+    echo "/usr/bin/xmllint is missing - install it to validate command output"
+fi
```
Comment (kgaillot) [inline cts/cts-cli.in:1770]: We usually check $(which xmllint) to allow for local environment flexibility. Using a variable for the xmllint location would help avoid the previous problem too

**Review by kgaillot — COMMENTED** [review]
"one more thing"

**Comment by clumens** [comment]
Too many things going on here at once.  This push should look much better.

File: .travis.yml [hunk 5]
```diff
@@ -54,7 +54,7 @@ install:
       libqb-dev libcfg-dev libcmap-dev libcorosync-common-dev libcpg-dev
       libquorum-dev libsam-dev libtotem-pg-dev libvotequorum-dev
       cluster-glue-dev resource-agents
- - test $MAINT_EXTRA -eq 0 || sudo apt-get install -qq libxml2-utils xsltproc
+      libxml2-utils xsltproc
```
Comment (kgaillot) [inline .travis.yml:57]: I think we lose travis-ci.org tomorrow so this may be moot. I haven't decided whether it's worth the hassle to move to .com ...

**Comment by kgaillot** [comment]
Whoops, this doesn't work when run from an installed deployment

**Comment by clumens** [comment]
I'll take a look.

**Pull request merged**